### PR TITLE
fix(snapshot): refactoring how snapshot sizes were computed

### PR DESCRIPTION
**Pull request by paologallinaharbur**
#### Description of the changes

Refactoring snapshot sizes to include vmdk files

Based on the govc [implementation](https://github.com/vmware/govmomi/blob/88629ae3fe3b069cbc3ff9f86e864b7e331bdc1c/object/virtual_machine.go#L692)

### CODE REVIEW

File: internal/process/snapshot.go [hunk 1]
```diff
@@ -4,109 +4,205 @@
...
-	"github.com/vmware/govmomi/vim25/types"
 
-	"github.com/newrelic/infra-integrations-sdk/data/metric"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware/govmomi/vim25/types"
 )
 
+const invalidFile = -1
+
+type snapshotProcessor struct {
```
Comment (gsanchezgavier) [inline internal/process/snapshot.go:19]: just if make sense, could you document really shortly how snapshots are orginized , or add some link ?

File: internal/process/snapshot.go [hunk 2]
```diff
@@ -4,109 +4,205 @@
...
-	if ex != nil {
-		for _, exFile := range ex.File {
-			if exFile.Type == "snapshotData" {
-				findSnapshotAndUpdateInfoData(ex.Snapshot, exFile, info)
-			} else if exFile.Type == "snapshotMemory" {
-				findSnapshotAndUpdateInfoMemory(ex.Snapshot, exFile, info)
-			} else if exFile.Type == "suspendMemory" {
-				suspendMemory += exFile.Size
-				suspendMemoryUnique += exFile.UniqueSize
+func (sp snapshotProcessor) processSnapshotTree(parentSnap *types.ManagedObjectReference, st []types.VirtualMachineSnapshotTree) {
```
Comment (gsanchezgavier) [inline internal/process/snapshot.go:28]: `parentSnapshot` , `snapshotTree` perhaps?

File: internal/process/snapshot.go [hunk 3]
```diff
@@ -4,109 +4,205 @@
...
+	return
+}
+
+// Function logic taken from the govc implementation
+//
+// SnapshotSize calculates the size of a given snapshot in bytes.
+// If the snapshot is current one, the disk files not associated with any parent snapshot are
+// included in size calculations. This allows for measuring and including the
+// growth from the last fixed snapshot to the present state.
+func (sp snapshotProcessor) snapshotSize(parent *types.ManagedObjectReference, snapshotRef types.ManagedObjectReference) {
```
Comment (gsanchezgavier) [inline internal/process/snapshot.go:44]: Perhaps we could chop this function and split it so is easier to follow.

File: internal/process/snapshot.go [hunk 4]
```diff
@@ -21,9 +22,62 @@ type snapshotProcessor struct {
...
+func buildFileKeyMap(filesInfo []types.VirtualMachineFileLayoutExFileInfo) map[int32]types.VirtualMachineFileLayoutExFileInfo {
+	fileKeyMap := map[int32]types.VirtualMachineFileLayoutExFileInfo{}
+	for _, file := range filesInfo {
+		fileKeyMap[file.Key] = file
+	}
+	return fileKeyMap
+}
+
+func buildSnapshotFileKeyMap(layout []types.VirtualMachineFileLayoutExSnapshotLayout) (map[types.ManagedObjectReference]snapshotInfo, []int32) {
+	var allFiles []int32
```
Comment (paologallinaharbur) [inline internal/process/snapshot.go:66]: changing it to allSnapshotsFiles

File: internal/process/snapshot.go [hunk 5]
```diff
@@ -38,24 +92,34 @@ func (sp snapshotProcessor) processSnapshotTree(parentSnapshot *types.ManagedObj
...
+		if file, ok := sp.filesInfoByKey[files.memory]; ok {
 			sp.results[snapshotRef].totalUniqueMemoryInDisk = file.UniqueSize
 			sp.results[snapshotRef].totalMemoryInDisk = file.Size
 			sp.results[snapshotRef].datastorePathMemory = file.Name
 		}
 	}
 }
 
+func (sp snapshotProcessor) isCurrent(snapshot types.ManagedObjectReference) bool {
+	var isCurrent bool
```
Comment (paologallinaharbur) [inline internal/process/snapshot.go:112]: sinplifying

File: internal/process/snapshot.go (hunk 4 above)
Comment (paologallinaharbur) [reply internal/process/snapshot.go:66]: I am improving other names also

File: internal/process/snapshot.go [hunk 6]
```diff
@@ -4,109 +4,283 @@
...
+			result = append(result, link.FileKey...)
+		}
+	}
+
+	return result
+}
+
+// removeKey is a helper function for removing a specific file key from a list
+// of keys associated with disks attached to a virtual machine.
+func removeKey(l *[]int32, key int32) {
```
[the remark this answers is not in the document]
Comment (paologallinaharbur) [reply internal/process/snapshot.go:236]: I had implemented it differently before, https://github.com/newrelic/nri-vsphere/pull/165/commits/5af9a3b0290247cc0241b38c681898d7d0bd59e5

However, it is the very same implementation of govc. I'll create a couple of tests to cover edge cases!

**Review by gsanchezgavier — APPROVED** [review]
thanks for taking care, and thanks for adding integration tests in case we need to improve performance in the future
